src/openrct2/world/Surface.cpp

```diff
@@ -9,49 +9,209 @@
+#include "../scenario/Scenario.h"
+#include "Location.hpp"
+#include "Map.h"
+
+uint32_t SurfaceElement::GetSurfaceStyle() const
 {
-    int32_t terrain = (element->properties.surface.terrain >> 5) & 7;
-    if (element->type & 1)
-        terrain |= (1 << 3);
-    return terrain;
+    uint32_t retVal = (terrain >> 5) & 7;
+    if (type & 1)
+        retVal |= (1 << 3);
+    return retVal;
 }
 
-int32_t surface_get_terrain_edge(const rct_tile_element* element)
+uint32_t SurfaceElement::GetEdgeStyle() const
 {
-    int32_t terrain_edge = (element->properties.surface.slope >> 5) & 7;
-    if (element->type & 128)
+    uint32_t terrain_edge = (slope >> 5) & 7;
+    if (type & 128)
         terrain_edge |= (1 << 3);
     return terrain_edge;
 }
 
-void surface_set_terrain(rct_tile_element* element, int32_t terrain)
+void SurfaceElement::SetSurfaceStyle(uint32_t newStyle)
 {
     // Bit 3 for terrain is stored in element.type bit 0
-    if (terrain & 8)
-        element->type |= 1;
+    if (newStyle & 8)
+        type |= 1;
     else
-        element->type &= ~1;
+        type &= ~1;
 
     // Bits 0, 1, 2 for terrain are stored in element.terrain bit 5, 6, 7
-    element->properties.surface.terrain &= ~0xE0;
-    element->properties.surface.terrain |= (terrain & 7) << 5;
+    terrain &= ~0xE0;
+    terrain |= (newStyle & 7) << 5;
 }
 
-void surface_set_terrain_edge(rct_tile_element* element, int32_t terrain)
+void SurfaceElement::SetEdgeStyle(uint32_t newStyle)
 {
     // Bit 3 for terrain is stored in element.type bit 7
-    if (terrain & 8)
-        element->type |= 128;
+    if (newStyle & 8)
+        type |= 128;
     else
-        element->type &= ~128;
+        type &= ~128;
 
     // Bits 0, 1, 2 for terrain are stored in element.slope bit 5, 6, 7
-    element->properties.surface.slope &= ~TILE_ELEMENT_SURFACE_EDGE_STYLE_MASK;
-    element->properties.surface.slope |= (terrain & 7) << 5;
+    slope &= ~TILE_ELEMENT_SURFACE_EDGE_STYLE_MASK;
+    slope |= (newStyle & 7) << 5;
+}
+
+uint32_t SurfaceElement::GetWaterHeight() const
+{
+    return terrain & TILE_ELEMENT_SURFACE_WATER_HEIGHT_MASK;
+}
+
+void SurfaceElement::SetWaterHeight(uint32_t newWaterHeight)
+{
+    newWaterHeight &= 0x1F;
+    terrain &= ~TILE_ELEMENT_SURFACE_WATER_HEIGHT_MASK;
+    terrain |= newWaterHeight;
+}
+
+uint8_t SurfaceElement::GetGrassLength() const
+{
+    return grass_length;
+}
+
+void SurfaceElement::SetGrassLength(uint8_t newLength)
+{
+    grass_length = newLength;
+}
+
+void SurfaceElement::SetGrassLengthAndInvalidate(uint8_t length, CoordsXY coords)
+{
+    uint8_t oldLength = grass_length & 0x7;
+    uint8_t newLength = length & 0x7;
+
+    grass_length = length;
+
+    if (newLength == oldLength)
+    {
+        return;
+    }
+
+    // If the new grass length won't result in an actual visual change
+    // then skip invalidating the tile, no point
+    if (((oldLength > 0 && oldLength < 4) && (newLength > 0 && newLength < 4))
```

Comment:
This might be easier to read if all the comparisons go low-to-high:

`(0 < oldLength && oldLength < 4) && (0 < newLength && newLength < 4)` etc.

(this is a drive-by comment that you're free to ignore; just noticed it as I was idly looking through the change out of curiosity)